src/dkg/round1.rs

```diff
@@ -121,11 +129,114 @@ pub fn import_secret_package(
     SerializableSecretPackage::deserialize_from(&serialized[..]).map(|pkg| pkg.into())
 }
 
+#[must_use]
+fn input_checksum<I>(min_signers: u16, signing_participants: &[I]) -> Checksum
+where
+    I: Borrow<Identity>,
+{
+    let mut hasher = SipHasher24::new();
+
+    hasher.write(&min_signers.to_le_bytes());
+
+    let mut signing_participants = signing_participants
+        .iter()
+        .map(Borrow::borrow)
+        .collect::<Vec<_>>();
+    signing_participants.sort_unstable();
+    signing_participants.dedup();
+
+    for id in signing_participants {
+        hasher.write(&id.serialize());
+    }
+
+    hasher.finish()
+}
+
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub struct Package {
+    identity: Identity,
+    frost_package: FrostPackage,
+    group_secret_key_part: [u8; 32],
+    checksum: Checksum,
+}
+
+impl Package {
+    pub fn new(
+        identity: Identity,
+        min_signers: u16,
+        signing_participants: &[Identity],
+        frost_package: FrostPackage,
+        group_secret_key_part: [u8; 32],
+    ) -> Self {
+        let checksum = input_checksum(min_signers, signing_participants);
+
+        Package {
+            identity,
+            frost_package,
+            group_secret_key_part,
+            checksum,
+        }
+    }
+
+    pub fn identity(&self) -> &Identity {
+        &self.identity
+    }
+
+    pub fn frost_package(&self) -> &FrostPackage {
+        &self.frost_package
+    }
+
+    pub fn group_secret_key_part(&self) -> [u8; 32] {
+        self.group_secret_key_part
+    }
+
+    pub fn checksum(&self) -> Checksum {
+        self.checksum
+    }
+
+    pub fn serialize(&self) -> io::Result<Vec<u8>> {
+        let mut buf = Vec::new();
+        self.serialize_into(&mut buf)?;
+        Ok(buf)
+    }
+
+    pub fn serialize_into<W: io::Write>(&self, mut writer: W) -> io::Result<()> {
+        self.identity.serialize_into(&mut writer)?;
+        writer.write_all(&self.frost_package.serialize().map_err(io::Error::other)?)?;
+        writer.write_all(&self.group_secret_key_part)?;
+        writer.write_all(&self.checksum.to_le_bytes())?;
+        Ok(())
+    }
+
+    pub fn deserialize_from<R: io::Read>(mut reader: R) -> io::Result<Self> {
+        let identity = Identity::deserialize_from(&mut reader).map_err(io::Error::other)?;
+
+        let frost_package = read_variable_length_bytes(&mut reader)?;
```

Comment:
I don't think this will work, given that there's no corresponding `write_variable_length_bytes` call above

Reply:
It doesn't work 😅 

I just changed the serialization to use `write_variable_length_bytes` and added a round trip serialization test